mods/explorer-ctrln-newfile.wh.cpp

```diff
@@ -0,0 +1,435 @@
+
+    HANDLE readyEvent = CreateEventW(nullptr, TRUE, FALSE, nullptr);
+    if (!readyEvent) return FALSE;
+
+    HANDLE hThread = CreateThread(nullptr, 0, HookThread, readyEvent, CREATE_SUSPENDED, &g_hookThreadId);
+    if (!hThread) {
+        CloseHandle(readyEvent);
+        return FALSE;
+    }
+
+    SetThreadPriority(hThread, THREAD_PRIORITY_ABOVE_NORMAL);
+    ResumeThread(hThread);
+
+    WaitForSingleObject(readyEvent, INFINITE);
+    CloseHandle(readyEvent);
+
+    if (!g_lowLevelHook) {
+        Wh_Log(L"[Ctrln] SetWindowsHookEx failed.");
+        WaitForSingleObject(hThread, INFINITE);
+        CloseHandle(hThread);
+        g_hookThreadId = 0;
+        return FALSE;
+    }
+    g_hookThread = hThread;
+    return TRUE;
+}
+
+static void KeybdHook_Exit() {
+    HANDLE hThread = (HANDLE)InterlockedExchangePointer((PVOID*)&g_hookThread, nullptr);
+    if (!hThread) return;
+
+    if (g_hookThreadId) PostThreadMessageW(g_hookThreadId, WM_APP, 0, 0);
+    WaitForSingleObject(hThread, INFINITE);
+    CloseHandle(hThread);
+    g_hookThreadId = 0;
+    g_lowLevelHook = nullptr;
+}
+
+static DWORD WINAPI HookThread(void* pParameter) {
+    HANDLE readyEvent = (HANDLE)pParameter;
+    MSG msg;
+
+    PeekMessageW(&msg, nullptr, WM_USER, WM_USER, PM_NOREMOVE);  // create queue
+    g_lowLevelHook = SetWindowsHookExW(WH_KEYBOARD_LL, LowLevelKeybdProc, HINST_THISCOMPONENT, 0);
+    SetEvent(readyEvent);
+    if (!g_lowLevelHook) return 0;
+
+    while (true) {
+        BOOL bRet = GetMessageW(&msg, nullptr, 0, 0);
+        if (bRet <= 0) break;
+
+        if (msg.hwnd == nullptr) {
+            if (msg.message == WM_APP) {
+                PostQuitMessage(0);
+                continue;
+            }
+            if (msg.message == WM_APP + 1) {
+                PerformNewFileAction();
+                continue;
+            }
+        }
+
+        TranslateMessage(&msg);
+        DispatchMessageW(&msg);
+    }
+
+    UnhookWindowsHookEx(g_lowLevelHook);
+    g_lowLevelHook = nullptr;
+    return 0;
+}
+
+static LRESULT CALLBACK LowLevelKeybdProc(int nCode, WPARAM wParam, LPARAM lParam) {
+    if (nCode == HC_ACTION) {
+        const KBDLLHOOKSTRUCT* info = reinterpret_cast<KBDLLHOOKSTRUCT*>(lParam);
+        bool isKeyDown = (wParam == WM_KEYDOWN || wParam == WM_SYSKEYDOWN);
+        bool isKeyUp = (wParam == WM_KEYUP || wParam == WM_SYSKEYUP);
+
+        if (info && isKeyUp && info->vkCode == 'N') g_nKeyDown = false;
+        if (info && isKeyUp && (info->vkCode == VK_LCONTROL || info->vkCode == VK_RCONTROL)) g_nKeyDown = false;
+
+        if (isKeyDown && info && info->vkCode == 'N' && !g_nKeyDown) {
+            const bool ctrl = (GetAsyncKeyState(VK_CONTROL) & 0x8000) != 0;
+            const bool shift = (GetAsyncKeyState(VK_SHIFT) & 0x8000) != 0;
+            if (ctrl && !shift) {
+                HWND fg = GetForegroundWindow();
+                if (IsExplorerTopLevel(fg) || fg == GetShellWindow()) {
```

Comment:
There might be several explorer.exe instances. It's better to make sure that the window is part of the current process with `GetWindowThreadProcessId`.